aten/src/ATen/native/Histogram.cpp

```diff
@@ -0,0 +1,133 @@
+#include <ATen/ATen.h>
+#include <ATen/Dispatch.h>
+#include <ATen/NativeFunctions.h>
+
+#include <ATen/native/Histogram.h>
+
+#include <tuple>
+
+namespace at { namespace native {
+
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+DEFINE_DISPATCH(histogram_stub);
+
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+DEFINE_DISPATCH(histogram_linear_stub);
+
+namespace {
+
+inline void histogram_pre_check(const Tensor& input, const Tensor& bins, const c10::optional<Tensor>& weight,
+    bool density, const Tensor& hist, const Tensor& bin_edges) {
+  TORCH_CHECK(bins.device() == input.device(), "bins and input value tensors should have same device type, ",
+    "but we got bins tensor device type ", bins.device(), " and input value tensor device type ", input.device());
+
+  TORCH_CHECK(bins.dim() == 1, "bins tensor should have dimension 1, but got ", bins.dim(), " dimension");
```

Comment:
This is a great error message, but let's start it with the name of the operation:

"torch.histogram(): bins tensors must have..."

What's great about the current error message is that it clarifies what's expected, provides the actual value, and is consistent with our error message pattern in other operations (minus the name of the op)

Reply:
Separate question (which I'll inline later): why is bins a tensor?

Reply:
The user-facing "bins" argument accepts either a tensor explicitly defining bin edges or an integer defining the number of bins. In the latter case, a tensor containing the implictly defined edges is constructed before we get to these checks. I've updated the variable naming within this function; it may be a bit clearer now.